lite/src/verifier.rs

```diff
@@ -0,0 +1,181 @@
+use crate::types::*;
+
+/// Returns an error if the header has expired according to the given
+/// trusting_period and current time. If so, the verifier must be reset subjectively.
+/// NOTE: this doesn't belong here. It should be called by something that handles whether to trust
+/// a verifieds commit. Verified here is really just about the header/commit/validators. Time is an
+/// external concern :)
+fn expired<H>(last_header: &H, trusting_period: Time, now: Time) -> Result<(), Error>
+where
+    H: Header,
+{
+    if last_header.bft_time() + trusting_period < now {
+        return Err(Error::Expired);
+    }
+    Ok(())
+}
+
+fn validate_next_vals<H, V>(header: H, next_vals: &V) -> Result<(), Error>
+where
+    H: Header,
+    V: ValidatorSet,
+{
+    // ensure the next validators in the header matches what was supplied.
+    if header.next_validators_hash() != next_vals.hash() {
+        return Err(Error::InvalidNextValidatorSet);
+    }
+
+    Ok(())
+}
+
+// Validate the validators and commit against the header.
+fn validate_vals_and_commit<H, V, C>(header: H, commit: &C, vals: &V) -> Result<(), Error>
+where
+    H: Header,
+    V: ValidatorSet,
+    C: Commit,
+{
+    // ensure the validators in the header matches what we expect from our state.
+    if header.validators_hash() != vals.hash() {
+        return Err(Error::InvalidValidatorSet);
+    }
+
+    // ensure the commit matches the header.
+    if header.hash() != commit.header_hash() {
+        return Err(Error::InvalidCommitValue);
+    }
+
+    Ok(())
+}
+
+/// Verify the commit is valid from the given validators for the header.
+pub fn verify<H, V, C>(header: H, commit: C, validators: V) -> Result<(), Error>
+where
+    H: Header,
+    V: ValidatorSet,
+    C: Commit,
+{
+    if let Err(e) = validate_vals_and_commit(header, &commit, &validators) {
+        return Err(e);
+    }
+
+    // ensure that +2/3 validators signed correctly
+    verify_commit_full(&validators, commit)
+}
+
+/// Verify the commit is trusted according to the last validators and is valid
+/// from the current validators for the header.
+pub fn verify_trusting<H, V, C>(
+    header: H,
+    commit: C,
+    last_validators: V,
+    validators: V,
+) -> Result<(), Error>
+where
+    H: Header,
+    V: ValidatorSetLookup,
+    C: Commit,
+{
+    if let Err(e) = validate_vals_and_commit(header, &commit, &validators) {
```

Comment:
```suggestion
    if let Err(e) = validate(header, &commit, &validators) {
```
and remove call to `verify_commit_full()` below?

Reply:
Why these changes ?

Reply:
`validate()` already calls `validate_vals_and_commit()` and `verify_commit_full()`. 
Also to reflect that all checks done by sequential client are also done by the skipping one.

Reply:
Ah, you mean `verify` not `validate` ?

I agree this might be nicer, but also there's a performance trade off - for the skipping ("trusting") case, I don't think we want to bother verifying the full commit until we verify that +1/3 of the old validator set signed. If we had it just call `verify` first, then we'd verify the full commit before we even discover if we can trust it according to the trusting method.

Reply:
> Ah, you mean `verify` not `validate` ?

Yes, sorry I meant `verify()`.

> I agree this might be nicer, but also there's a performance trade off - for the skipping ("trusting") case, I don't think we want to bother verifying the full commit until we verify that +1/3 of the old validator set signed. If we had it just call `verify` first, then we'd verify the full commit before we even discover if we can trust it according to the trusting method.

I think the complexity is same for both. But to address your concern, can't we call `verify_commit_trusting` follow by `verify`?